src/server/mysql_conn.cpp

```diff
@@ -0,0 +1,469 @@
+		account_->set_connect_option(mysql_option::MYSQL_SET_CHARSET_NAME, std::string("utf8mb4"));
+		account_->set_schema(c["db_name"].str());
+		connection_ = mariadb::connection::create(account_);
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Failed to connect to the database!", e);
+	}
+
+	try
+	{
+		mp_mod_group_ = std::stoi(c["mp_mod_group"].str());
+	}
+	catch(...)
+	{
+		ERR_SQL << "Failed to convert the mp_mod_group value of '" << c["mp_mod_group"].str() << "' into an int!  Defaulting to " << mp_mod_group_ << "." << std::endl;
+	}
+}
+
+void mysql_conn::log_sql_exception(const std::string& text, const mariadb::exception::base& e)
+{
+	ERR_SQL << text << '\n'
+			<< "what: " << e.what() << '\n'
+			<< "error id: " << e.error_id() << std::endl;
+}
+
+//
+// queries
+//
+std::string mysql_conn::get_uuid()
+{
+	try
+	{
+		return get_single_string("SELECT UUID()");
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Could not retrieve a UUID!", e);
+		return "";
+	}
+}
+
+std::string mysql_conn::get_tournaments()
+{
+	if(db_tournament_query_ == "")
+	{
+		return "";
+	}
+
+	try
+	{
+		std::string text;
+		for(const auto& row : get_string_data(db_tournament_query_))
+		{
+			text += "\nThe tournament "+row.at("TITLE")+" is "+row.at("STATUS")+". More information can be found at "+row.at("URL");
+		}
+		return text;
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Could not retrieve the tournaments!", e);
+		return "";
+	}
+}
+
+bool mysql_conn::user_exists(const std::string& name)
+{
+	try
+	{
+		return exists("SELECT 1 FROM `"+db_users_table_+"` WHERE UPPER(username)=UPPER(?)", name);
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Unable to check if user row for '"+name+"' exists!", e);
+		return false;
+	}
+}
+
+bool mysql_conn::extra_row_exists(const std::string& name)
+{
+	try
+	{
+		return exists("SELECT 1 FROM `"+db_extra_table_+"` WHERE UPPER(username)=UPPER(?)", name);
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Unable to check if extra row for '"+name+"' exists!", e);
+		return false;
+	}
+}
+
+bool mysql_conn::is_user_in_group(const std::string& name, int group_id)
+{
+	try
+	{
+		return exists("SELECT 1 FROM `"+db_users_table_+"` u, `"+db_user_group_table_+"` ug WHERE UPPER(u.username)=UPPER(?) AND u.USER_ID = ug.USER_ID AND ug.GROUP_ID = ?", name, group_id);
+	}
+	catch(const mariadb::exception::base& e)
+	{
+		log_sql_exception("Unable to check if the user '"+name+"' is in group '"+std::to_string(group_id)+"'!", e);
+		return false;
+	}
+}
+
+// NOTE: A ban end time of 0 is a permanent ban.
+bool mysql_conn::ip_is_banned(const std::string& ip)
+{
+	try
+	{
+		return exists("SELECT 1 FROM `"+db_banlist_table_+"` WHERE UPPER(ban_ip) = UPPER(?) AND ban_exclude = 0 AND (ban_end = 0 OR ban_end >="+std::to_string(std::time(nullptr)) + ")", ip);
```

Comment:
Is there a reason to not use the prepared statement syntax for the ban time as well? (Same thing for the two other similar cases below.)

Reply:
Not especially.